### PR TITLE
Alerting: Persist notification log and silences to the database

### PR DESCRIPTION
**Pull request by gotjosh**
This removes the dependency of having a persistent disk to run Grafana alerting. Instead of regularly flushing the notification log and silences to disk we now flush the binary content of those files to the database encoded as a base64 string.

There's still a couple of things left to figure out:

- [x] how does this work with multiple replicas - do I need to store each log independently per replica? This only matter when clustering and if its position is > 0. I'll take into account as part of #37920
- [x] Provisioning of the kvstore - is this the way wire works? 
- [x] Update documentation
- [x] Does this needs to be behind a flag? Probably not.
- [x] Figure out if we need to timeout while flushing to the database. Given this runs in a separate goroutine I believe this is non-blocking but need to double-check if we hold one of the internal locks for too long.

### CODE REVIEW

File: docs/sources/alerting/unified-alerting/_index.md [hunk 1]
```diff
@@ -47,14 +47,14 @@ Alerting rules can only query backend data sources with alerting enabled:
 
 The alerting engine publishes some internal metrics about itself. You can read more about how Grafana publishes [internal metrics]({{< relref "../../administration/view-server/internal-metrics.md" >}}).
 
-| Metric Name                                 | Type      | Description                                                                              |
```
Comment (gotjosh) [inline docs/sources/alerting/unified-alerting/_index.md:50]: Unrelated but found these as I looked through the documentation.

File: pkg/services/ngalert/notifier/file_store.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+// If there is a file in the database, it decodes it and writes to disk for Alertmanager consumption.
+func (fs *FileStore) FilepathFor(filename string) (string, error) {
+	// If a file is already present, we'll use that one and eventually save it to the database.
+	// We don't need to do anything else.
+	if fs.IsExists(filename) {
+		return fs.pathFor(filename), nil
+	}
+
+	// Then, let's attempt to read it from the database.
+	content, exists, err := fs.kv.Get(context.Background(), filename)
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/file_store.go:49]: There's probably some merit in timing out on this call - not doing so poses the risk of sync taking too long. Doing so poses the risk of eventually overriding a configuration in the database...

Of the two, I chose the former until we get some customer feedback - I'm happy to be swayed the other way.

File: pkg/services/ngalert/notifier/file_store.go [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	if err := fs.WriteFileToDisk(filename, bytes); err != nil {
+		return "", fmt.Errorf("error writing file %s: %w", filename, err)
+	}
+
+	return fs.pathFor(filename), err
+}
+
+// Persist takes care of persisting the binary representation of internal state to the database as a base64 encoded string.
+func (fs *FileStore) Persist(filename string, st State) (int64, error) {
```
Comment (gerobinson) [inline pkg/services/ngalert/notifier/file_store.go:73]: I think `context.Context` should be the first argument here too?

File: pkg/services/ngalert/notifier/file_store_test.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	"context"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+func TestFileStore_FilepathFort(t *testing.T) {
```
Comment (gerobinson) [inline pkg/services/ngalert/notifier/file_store_test.go:13]: Spelling mistake in `FilepathFor`

**Review by grobinson-grafana — APPROVED** [review]

**Review by grobinson-grafana — APPROVED** [review]

**Pull request merged**